test/devapps/device_code_flow_sample.go

```diff
@@ -16,51 +17,45 @@ func deviceCodeCallback(deviceCodeResult msal.DeviceCodeResultProvider) {
 	log.Infof(deviceCodeResult.GetMessage())
 }
 
-func tryDeviceCodeFlow(publicClientApp *msal.PublicClientApplication) {
-	cancelTimeout := 100 //Change this for cancel timeout
-	cancelCtx, cancelFunc := context.WithTimeout(context.Background(), time.Duration(cancelTimeout)*time.Second)
-	defer cancelFunc()
-	deviceCodeParams := msal.CreateAcquireTokenDeviceCodeParameters(cancelCtx, config.Scopes, deviceCodeCallback)
-	resultChannel := make(chan msal.AuthenticationResultProvider)
-	errChannel := make(chan error)
-	go func() {
-		result, err := publicClientApp.AcquireTokenByDeviceCode(deviceCodeParams)
-		errChannel <- err
-		resultChannel <- result
-	}()
-	err = <-errChannel
-	if err != nil {
-		log.Fatal(err)
-	}
-	result := <-resultChannel
-	fmt.Println("Access token is " + result.GetAccessToken())
-}
-
 func acquireTokenDeviceCode() {
 	config := createConfig("config.json")
-	publicClientApp, err := msal.CreatePublicClientApplication(config.ClientID, config.Authority)
-	if err != nil {
-		log.Fatal(err)
-	}
-	publicClientApp.SetCacheAccessor(cacheAccessor)
-	var userAccount msal.AccountProvider
-	accounts := publicClientApp.GetAccounts()
+	// create a PublicClientApplication with a  custom cache accessor
+	options := msal.DefaultPublicClientApplicationOptions()
+	options.Accessor = cacheAccessor
+	publicClientApp := msal.NewPublicClientApplication(config.ClientID, config.Authority, &options)
+
+	// look in the cache to see if the account to use has been cached
+	var userAccount *msalbase.Account
+	accounts := publicClientApp.Accounts()
 	for _, account := range accounts {
-		if account.GetUsername() == config.Username {
+		if account.PreferredUsername == config.Username {
 			userAccount = account
 		}
 	}
-	if userAccount == nil {
-		log.Info("No valid account found")
-		tryDeviceCodeFlow(publicClientApp)
-	} else {
-		silentParams := msal.CreateAcquireTokenSilentParametersWithAccount(config.Scopes, userAccount)
-		result, err := publicClientApp.AcquireTokenSilent(silentParams)
+	var result *msalbase.AuthenticationResult
+	var err error
+	if userAccount != nil {
```

Comment:
Do we need to test if the `userAccount` is not `nil`? couldn't just `AcquireTokenSilent` return an error if the `userAccount` is `nil`? (Other MSALs do that)

Reply:
With the current implementation, a non-nil error is ambiguous; it means that either the user account wasn't found or that something else failed (e.g. perhaps there was a transient IO failure making the network request).

I think we need to decide on the semantics we want here, and also look at what kind of error(s) other implementations expose here.  I've opened https://github.com/AzureAD/microsoft-authentication-library-for-go/issues/70 to track standardizing on error types.